libraries/tock-register-interface/README.md

```diff
@@ -3,130 +3,107 @@
-registers must be explicitly annotated with an offset and gap identifier (by
-convention using a field named `_reservedN`), but without a type. The macro will
-then automatically take care of calculating the gap size and inserting a
-suitable filler struct. The end of the struct is marked with its size and the
-`@END` keyword, effectively pointing to the offset immediately past the list of
-registers.
+## Defining peripherals
 
-```rust
-use tock_registers::registers::{ReadOnly, ReadWrite, WriteOnly};
+An MMIO peripheral is defined using the `peripheral!` macro:
 
-register_structs! {
+```rust
+tock_registers::peripheral! {
     Registers {
-        // Control register: read-write
-        // The 'Control' parameter constrains this register to only use fields from
-        // a certain group (defined below in the bitfields section).
-        (0x000 => cr: ReadWrite<u8, Control::Register>),
-
-        // Status register: read-only
-        (0x001 => s: ReadOnly<u8, Status::Register>),
-
-        // Registers can be bytes, halfwords, or words:
-        // Note that the second type parameter can be omitted, meaning that there
-        // are no bitfields defined for these registers.
-        (0x002 => byte0: ReadWrite<u8>),
-        (0x003 => byte1: ReadWrite<u8>),
-        (0x004 => short: ReadWrite<u16>),
+        // Control register: read-write. `u8` specifies the data type of the
+        // memory-mapped register, and the `Control` parameter specifies how
+        // bitfields within the register should be accessed.
+        0x000 => cr: u8(Control::Register) { Read, Write },
+
+        // Status register: read-only. Registers can have documentation
+        // comments, like:
+        /// Status register
+        0x001 => s: u8(Status::Register) { Read },
+
+        // Registers can have different sizes. The bitfield parameter is
+        // optional:
+        0x002 => byte0: u8 { Read, Write },
+        0x003 => byte1: u8 { Read, Write },
+        0x004 => short: u16 { Read, Write },
 
         // Empty space between registers must be marked with a padding field,
-        // declared as follows. The length of this padding is automatically
-        // computed by the macro.
-        (0x006 => _reserved),
-        (0x008 => word: ReadWrite<u32>),
-
-        // The type for a register can be anything. Conveniently, you can use an
-        // array when there are a bunch of similar registers.
-        (0x00C => array: [ReadWrite<u32>; 4]),
-        (0x01C => ... ),
+        // specified by naming the field _ and omitting the operations list:
+        0x006 => _: u16,
 
-        // Etc.
+        // Padding's length may also be inferred by the macro by specifying _ as
+        // the data type:
+        0x008 => _: _,
 
-        // The end of the struct is marked as follows.
-        (0x100 => @END),
+        // If you have several adjacent registers, you can specify an array
+        // register:
+        0x00C => array: [u32; 4] { Read, Write },
```

Comment:
I f I would like to specify a specific bitfield type, would I write:

```rust
gpioPins: [u32(Gpio::Bank); 4] { Read, Write },
```

Reply:
I had a request to allow registers to be specified by type alias, i.e.:

```rust
peripheral! {
    GpioPort {
        0x0 => pins: GpioPins(GpioPin::Register) { Read, Write },
    }
}

type GpioPins = [u8; 16];
```

That requires the traits to infer the fact the register is an array from the type itself. As a result, `peripheral!` treats the type as an opaque value (a `syn::Type`) and does not try to determine whether a register is an array or scalar. That also rules out the `[u32(Gpio::Bank); 4]` syntax.

@kupiakos FYI

Reply:
I was thinking more like this syntax:

```rust
peripheral! {
    GpioPort {
        0x0 => pins: [Gpio::Bank; 4] { Read, Write },
    }
}
```

Is there a specific reason to provide the primitive type backing a bitfield inside of `peripheral!`? The bitfield register type should map to exactly one primitive type. I could see some value in documentation like the explicit offsets.

Reply:
> I was thinking more like this syntax:
> 
> ```rust
> peripheral! {
>     GpioPort {
>         0x0 => pins: [Gpio::Bank; 4] { Read, Write },
>     }
> }
> ```
> 
> Is there a specific reason to provide the primitive type backing a bitfield inside of `peripheral!`? The bitfield register type should map to exactly one primitive type. I could see some value in documentation like the explicit offsets.

I would love for users to only have to specify the bitfield type and have the macro/registers infer it. However, currently bitfield types implement the [`RegisterLongName`](https://github.com/tock/tock/blob/91e78eaa847bfb1d3a4db5c7139b034bf7e9a481/libraries/tock-register-interface/src/lib.rs#L132) trait, which doesn't indicate the underlying primitive type that it applies to. I would love to change the trait to:

```rust
pub trait RegisterLongName {
    // The underlying data type of this register (the u8/u16/u32/etc
    // passed over the MMIO bus).
    type Primitive: UIntLike;
}
```

However, many (most?) registers use the following `RegisterLongName` impl:

```rust
impl RegisterLongName for () {}
```

regardless of the primitive type. So adding `Primitive` would be a massively breaking change.

Instead, we could add a new trait that contains both the bitfield implementation and primitive, but that would result in sizable changes to the `register_bitfield!` infrastructure. I have two reasons for not implementing those changes:

1. I haven't used register bitfields enough to be confident implementing a v2 of it.
2. I'm on a deadline to implement these changes which does not support rewriting `register_bitfield!`.

> Another alternative syntax could be:
> 
> ```rust
> peripheral! {
>     GpioPort {
>         0x0 => pins: [u32; 4](GpioPin::Register) { Read, Write },
>     }
> }
> ```

This is the syntax I envisioned using; I added an example to the README. Definitely changeable though.

Reply:
I figured out how to make `peripheral!` work without requiring the user to specify both the value type and the bitfield type and also without totally rewriting the bitfield stuff. This leads to the following syntax:

```rust
use tock_registers::{Aliased, peripheral, register_bitfields};
register_bitfields![u32,
    Control [ ... ]
    Pin [ ... ]
    Status [ ... ]
    Trigger [ ... ]
];
peripheral! {
    Foo {
        0x0 => control: Control::Register { Write },

        // Array registers can have bitfields as well. The bitfield applies to
        // each register in the array.
        0x4 => pins: [Pin::Register; 4] { Read, Write },

        // Registers can have different bitfields for read operations than for
        // write operations. In that case, specify the type as
        // Aliased<read bitfield, write bitfield>
        0x8 => aliased: Aliased<Status::Register, Trigger::Register> { Read, Write },

        // Array registers can also be aliased:
        0xc => motors: [Aliased<Status::Register, Trigger::Register>; 4] { Read, Write },
    }
}
```

The value type (`u32` for all the registers here) is taken from the `register_bitfields!` invocation (transmitted through some new traits).

I personally prefer this; I always thought that specifying both the value type and the bitfield is redundant. However, this was a pretty big change to how the macro works, and making involved deleting a bunch of complex code. Therefore I pushed it as a separate commit, so I can roll it back if you don't like the changes. I'm a little unsure of the change because I do not understand why the RegisterLongName and value type were specified separately in the first place.

I can't really make further progress on this PR until I know which way we're going on this issue (the design diverged pretty hard), so lets discuss this at this Friday's core WG call and reach a consensus on which way to go.

Oh, and feel free to register your opinion on this latest change by using thumbs up and thumbs down on this comment.

Reply:
I _think_ the reason to specify them separately in the old (well... current) syntax is because the width of the field determines the offset of the next field. While, in this new syntax, that's unnecessary since the offset of the next field is expressed explicitly. So this seems striclty better

Reply:
By "old" syntax, are you referring to the syntax that predates offset tests? i.e.:

```rust
#[repr(C)]
struct Registers {
    a: ReadWrite<u32, Ctrl::Register>,
    b: ReadWrite<u8, Status::Register>,
}
```

If so that makes sense to me!

Reply:
My opinions:

- The new syntax is strictly an improvement.
- It would be less confusing as a user to have the `::Register` type be a `#[repr(transparent)]` struct over the data type with the expected size, rather than with another ZST and associated type - especially when `u32` and `u8`, placed in the same location, are the actual primitive types being read.

Reply:
In today's core WG call, we decided to move forward with this new syntax. I will squash the two commits in this PR together and move forward with the new design.